libs/langchain/tests/unit_tests/agents/test_agent.py

```diff
@@ -626,6 +628,140 @@ def find_pet(pet: str) -> str:
 
 
+async def test_runnable_with_multi_action_per_step() -> None:
+    """Test an agent that can make multiple function calls at once."""
+    # Will alternate between responding with hello and goodbye
+    infinite_cycle = cycle(
+        [AIMessage(content="looking for pet..."), AIMessage(content="Found Pet")]
+    )
+    model = GenericFakeChatModel(messages=infinite_cycle)
+
+    template = ChatPromptTemplate.from_messages(
+        [("system", "You are Cat Agent 007"), ("human", "{question}")]
+    )
+
+    parser_responses = cycle(
+        [
+            [
+                AgentAction(
+                    tool="find_pet",
+                    tool_input={
+                        "pet": "cat",
+                    },
+                    log="find_pet()",
+                ),
+                AgentAction(
+                    tool="pet_pet",  # A function that allows you to pet the given pet.
+                    tool_input={
+                        "pet": "cat",
+                    },
+                    log="pet_pet()",
+                ),
+            ],
+            AgentFinish(
+                return_values={"foo": "meow"},
+                log="hard-coded-message",
+            ),
+        ],
+    )
+
+    def fake_parse(inputs: dict) -> Union[AgentFinish, AgentAction]:
+        """A parser."""
+        return cast(Union[AgentFinish, AgentAction], next(parser_responses))
+
+    @tool
+    def find_pet(pet: str) -> str:
+        """Find the given pet."""
+        if pet != "cat":
+            raise ValueError("Only cats allowed")
+        return "Spying from under the bed."
+
+    @tool
+    def pet_pet(pet: str) -> str:
+        """Pet the given pet."""
+        if pet != "cat":
+            raise ValueError("Only cats should be petted.")
+        return "purrrr"
+
+    agent = template | model | fake_parse
+    executor = AgentExecutor(agent=agent, tools=[find_pet])
+
+    # Invoke
+    result = executor.invoke({"question": "hello"})
+    assert result == {"foo": "meow", "question": "hello"}
+
+    # ainvoke
+    result = await executor.ainvoke({"question": "hello"})
+    assert result == {"foo": "meow", "question": "hello"}
+
+    # astream
+    results = [r async for r in executor.astream({"question": "hello"})]
+    assert results == [
+        {
+            "actions": [
+                AgentAction(
+                    tool="find_pet", tool_input={"pet": "cat"}, log="find_pet()"
+                )
+            ],
+            "messages": [AIMessage(content="find_pet()")],
+        },
+        {
+            "actions": [
+                AgentAction(tool="pet_pet", tool_input={"pet": "cat"}, log="pet_pet()")
+            ],
+            "messages": [AIMessage(content="pet_pet()")],
+        },
+        {
+            # By-default observation gets converted into human message.
+            "messages": [HumanMessage(content="Spying from under the bed.")],
```

Comment:
I didn't look into this message yet, looks reasonable if the assumption is that not all chat models support FunctionMessage